Sources/SwiftIfConfig/ConfiguredRegions.swift

```diff
@@ -56,58 +143,111 @@ fileprivate class ConfiguredRegionVisitor<Configuration: BuildConfiguration>: Sy
+    }
+
+    // Walk through the clauses to find the active one.
     var foundActive = false
     var syntaxErrorsAllowed = false
+    let outerState: IfConfigRegionState = inActiveRegion ? .active : .inactive
     for clause in node.clauses {
-      // If we haven't found the active clause yet, syntax errors are allowed
-      // depending on this clause.
-      if !foundActive {
-        syntaxErrorsAllowed =
-          clause.condition.map {
-            IfConfigClauseSyntax.syntaxErrorsAllowed($0).syntaxErrorsAllowed
-          } ?? false
-      }
+      let isActive: Bool
+      if let condition = clause.condition {
+        if !foundActive {
+          // Fold operators so we can evaluate this #if condition.
+          let (foldedCondition, foldDiagnostics) = IfConfigClauseSyntax.foldOperators(condition)
+          diagnostics.append(contentsOf: foldDiagnostics)
+
+          // In an active region, evaluate the condition to determine whether
+          // this clause is active. Otherwise, this clause is inactive.
+          // inactive.
+          if inActiveRegion {
+            let (thisIsActive, _, evalDiagnostics) = evaluateIfConfig(
+              condition: foldedCondition,
+              configuration: configuration
+            )
+            diagnostics.append(contentsOf: evalDiagnostics)
 
-      // If this is the active clause, record it and then recurse into the
-      // elements.
-      if clause == activeClause {
-        assert(inActiveRegion)
+            // Determine if there was an error that prevented us from
+            // evaluating the condition. If so, we'll allow syntax errors
+            // from here on out.
+            let hadError =
+              foldDiagnostics.contains { diag in
+                diag.diagMessage.severity == .error
+              }
+              || evalDiagnostics.contains { diag in
+                diag.diagMessage.severity == .error
+              }
 
-        regions.append((clause, .active))
+            if hadError {
+              isActive = false
+              syntaxErrorsAllowed = true
+            } else {
+              isActive = thisIsActive
 
-        if let elements = clause.elements {
-          walk(elements)
+              // Determine whether syntax errors are allowed.
+              syntaxErrorsAllowed = foldedCondition.allowsSyntaxErrorsFolded
+            }
+          } else {
+            isActive = false
+
+            // Determine whether syntax errors are allowed, even though we
+            // skipped evaluation of the actual condition.
+            syntaxErrorsAllowed = foldedCondition.allowsSyntaxErrorsFolded
+          }
+        } else {
+          // We already found an active condition, so this is inactive.
+          isActive = false
         }
+      } else {
+        // This is an #else. It's active if we haven't found an active clause
+        // yet and are in an active region.
+        isActive = !foundActive && inActiveRegion
```

Comment:
Do we still need to determine `syntaxErrorsAllowed` in that case. Eg. assuming that `DEBUG` is set, do we report `print debugDescription` as `syntaxErrorsAllowed` for the following

```swift
#if DEBUG
print(debugDescription)
#elseif compiler(>99.0)
print debugDescription
#endif
```

Reply:
The compiler doesn't determine `syntaxErrorsAllowed` here, and prints an error for `print debugDescription`. We could choose to deviate from the compiler here (and be more lax) if we want to.

Reply:
OK, that’s surprising to me but matching the compiler here for now makes sense for now to make switching between the two implementations easier. 

Should we gather these oddities that we find somewhere so we can consider fixing them in the future?

Reply:
It turns out that this behavior changed in the compiler very recently (https://github.com/swiftlang/swift/pull/74415). We could consider whether we need to fix the original issue in some other way.